src/registrar/models/domain.py

```diff
@@ -1648,6 +1667,105 @@ def _get_dnssec_data(self, response_extensions):
+        """
+        if fetch_hosts:
+            cleaned_hosts = cleaned["hosts"]
+            # Get all existing hosts from the database for this domain
+            existing_hosts_in_db = Host.objects.filter(domain=self)
+            # Identify hosts to delete
+            cleaned_host_names = set(cleaned_host["name"] for cleaned_host in cleaned_hosts)
+            hosts_to_delete_from_db = [
+                existing_host for existing_host in existing_hosts_in_db if existing_host.name not in cleaned_host_names
+            ]
+            # Delete hosts and their associated HostIP instances
+            for host_to_delete in hosts_to_delete_from_db:
+                # Delete associated HostIP instances
+                HostIP.objects.filter(host=host_to_delete).delete()
+                # Delete the host itself
+                host_to_delete.delete()
+            # Update or create Hosts and HostIPs
+            for cleaned_host in cleaned_hosts:
+                # Check if the cleaned_host already exists
+                host_in_db, host_created = Host.objects.get_or_create(domain=self, name=cleaned_host["name"])
+                # Get cleaned list of ips for update
+                cleaned_ips = cleaned_host["addrs"]
+                if not host_created:
+                    # Get all existing ips from the database for this host
+                    existing_ips_in_db = HostIP.objects.filter(host=host_in_db)
+                    # Identify IPs to delete
+                    ips_to_delete_from_db = [
+                        existing_ip for existing_ip in existing_ips_in_db if existing_ip.address not in cleaned_ips
+                    ]
+                    # Delete IPs
+                    for ip_to_delete in ips_to_delete_from_db:
+                        # Delete the ip
+                        ip_to_delete.delete()
+                # Update or create HostIP instances
+                for ip_address in cleaned_ips:
+                    HostIP.objects.get_or_create(address=ip_address, host=host_in_db)
+
+    def _update_security_contact_in_db(self, cleaned, fetch_contacts):
+        """Update security contact registry id in database if retrieved from registry.
+        If no value is retrieved from registry, set to empty string in db.
+
+        Parameters:
+            self: the domain to be updated with security from cleaned
+            cleaned: dict containing contact registry ids. Security contact is of type
+                PublicContact.ContactTypeChoices.SECURITY
+            fetch_contacts: boolean indicating whether or not fetch_contacts was called
+        """
+        if fetch_contacts:
+            cleaned_contacts = cleaned["contacts"]
+            security_contact_registry_id = ""
+            if cleaned_contacts[PublicContact.ContactTypeChoices.SECURITY]:
+                security_contact_registry_id = cleaned_contacts[PublicContact.ContactTypeChoices.SECURITY]
```

Comment:
(Nitpick/optional) cleaned_contacts[PublicContact.ContactTypeChoices.SECURITY] can be stored in a variable here since its reused, i.e. "security_contact"

Reply:
good suggestion, incorporated